[Carlintjy] iP

Carlintyj opened the pull request:
# Duke Carl

This ChatBot is able to
- Talk to you
- Be a ToDo List
- Keep track of events, deadlines and todo.

All you need to do is
1. Download the jar
2. java -jar

Snippet of my Duke Constructor
```java
public Duke(String filePath) {
        ui = new Ui();
        storage = new Storage(filePath);
        try {
            tasks = new TaskList(storage.load());
        } catch (DukeException e) {
            ui.showLoadingError();
            tasks = new TaskList();
        }
    }
```

Features to be completed
- [ x ] Add events, deadlines and todo
- [ ] GUI
- [ ] Checkstyle 🎉 

> Remember to not ~destroy~ the bot!

**Visit** the GitHub [here](https://github.com/Carlintyj/ip/tree/master)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Carlintyj committed:
Given chatbot a new name and implemented initial skeleton

Carlintyj committed:
Improve skeletal version by echoing commands by user

Carlintyj committed:
Added the ability to store whatever text entered by the user and display them back to the user when requested.

Carlintyj committed:
Added the ability to mark tasks as done and back to not done.

Carlintyj committed:
Added support for tracking three types of tasks

Carlintyj committed:
Test using I/O redirection technique

Carlintyj committed:
Added Error Handling

Carlintyj committed:
Added support for deleting tasks from the list.

Carlintyj committed:
Save tasks in the hard disk automatically whenever the task list changes.

Carlintyj committed:
Revert "Save tasks in the hard disk automatically whenever the task list changes."

This reverts commit c69d6b0133e26ecb05feb4d2a8bd9a96c4417471.

Carlintyj committed:
Revert "Revert "Save tasks in the hard disk automatically whenever the task list changes.""

This reverts commit 70384ae1d37fe6392808b7d584ec1ebc3c7cfc9a.

Carlintyj committed:
Store dates as LocalDate

Carlintyj committed:
Configured program into more OOP

Carlintyj committed:
Divide classes into packages

Carlintyj committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  Bump gradle and lib version
  Add Gradle support

Carlintyj committed:
Automate project builds using Gradle

Carlintyj committed:
Add JUnit tests

Carlintyj committed:
Package App as JAR file

Carlintyj committed:
Give users a way to find a task by searching for a keyword

Carlintyj committed:
Follow coding standard

Carlintyj committed:
JavaDocs for the Duke application

Carlintyj committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Give users a way to find a task by searching for a keyword

Carlintyj committed:
Uses checkstyle to detect coding violations and fix it.

Carlintyj committed:
Adds a GUI to the chatbot

Carlintyj committed:
Improves the GUI of chatbot

s-kybound commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,105 @@
+import duke.task.TaskList;
+import duke.task.Todo;
+import duke.ui.Ui;
+
```
Great job on importing each class explicitly instead of using *!
Now I know exactly which classes your program uses.

s-kybound commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,105 @@
+    @Override
+    public void execute(TaskList taskList, Ui ui, Storage storage) throws DukeException {
+        switch (type) {
+        case 'T':
```
Nicely done, following the correct indentation for switch case statements!

s-kybound commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    /**
+     * Executes the add task command, adding the specified task to the task list and updating the storage.
```
Love the Javadocs! They're well explained, illustrating what your code does well

s-kybound commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,27 @@
+ */
+public abstract class Command {
+
+    /**
```
Perhaps one suggestion would be to align the parameter descriptions? For example:

```suggestion
    /**
     * Executes the command.
     *
     * @param taskList       The list of tasks.
     * @param ui             The user interface to interact with the user.
     * @param storage        The storage to update the tasks in the file.
     * @throws DukeException If there's an error while parsing the user input or updating the storage.
     */
```

s-kybound commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,52 @@
+package duke;
+
+import java.util.concurrent.TimeUnit;
+
```
Great job logically separating the packages!

s-kybound commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,15 @@
+package duke;
+
+/**
```
Good job documenting the purpose of the classes!

s-kybound commented on src/main/java/duke/Launcher.java:
```diff
@@ -0,0 +1,13 @@
+import javafx.application.Application;
+
+/**
+ * A launcher class to workaround classpath issues.
```
Minor nitpick - is this supposed to be "work around" and not "workaround"?

s-kybound commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,55 @@
+/**
+ * Represents a command to delete tasks to the task list in the Duke application.
+ */
+public class DeleteCommand extends Command {
```
Good job using Egyptian bracing throughout the code!

s-kybound commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,55 @@
+            String deletedTask = taskList.getPrint(index);
+            taskList.deleteTask(index);
+            ui.sendMessage("Noted. I've removed this task:\n\t\t"
+                    + deletedTask
```
I love the way you break lines that are too long into several lines to adhere to the line limit and improve readability!

s-kybound commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,99 @@
+     * @throws DukeException If there's an error while parsing the user input or generating a Command.
+     */
+    public static Command parse(String userInput) throws DukeException {
+        if (userInput.equals("bye")) {
```
Perhaps a good improvement would be to refactor this as a switch case statement?

s-kybound commented on src/main/resources/images/DaUser.png:
Cute!

s-kybound commented on src/test/java/duke/command/AddCommandTest.java:
```diff
@@ -0,0 +1,58 @@
+
+public class AddCommandTest {
+    @Test
+    public void AddSingleTodo() throws DukeException {
```
Perhaps a good way to go forward would be to follow the naming convention for tests! For example, 

```
unitBeingTested_descriptionOfTestInputs_expectedOutcome
```

s-kybound reviewed (commented):
Overall, great coding standards! I can see the diligence in adhering to the course coding standard, and I have learned from your project :)

hyc17003 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,62 @@
+    /**
+     * Marks the task as done by setting the 'isDone' flag to true.
+     */
+    public void markAsDone() {
```
Maybe the method name could be changed to something like setDone to adhere to Java coding standards, as it is a setter method for a Boolean variable?
```suggestion
    public void setDone() {
```

hyc17003 commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,55 @@
+     */
+    @Override
+    public void execute(TaskList taskList, Ui ui, Storage storage) throws DukeException {
+        if (index >= taskList.size() || index < 0) {
```
I like that you handled this exception!

hyc17003 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,105 @@
+package duke.command;
```
I like the way you have segmented your code in separate command classes in the command package! It's very organised and readable.

hyc17003 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,105 @@
+ */
+public class AddCommand extends Command {
+    private String fullCommand;
+    private char type;
```
Perhaps this variable name could be clearer, something like taskType? Just to make it very obvious and prevent any ambiguity :-)

hyc17003 reviewed (commented):
LGTM, looks very organised and is easy to follow the logic! The graphics are a nice touch, they're really cute. Good job!

Carlintyj committed:
Command classes: complicated expressions and deep nestings.

Among various dimensions of code quality, such as run-time efficiency, security, and robustness, one of the most important is readability. (This makes it hard to read)

The common behaviors across classes cause code duplication.

Extracting the common behavior into a super class allows us to use
polymorphism when dealing with person attributes.

Different functions were created within the classes to allow better readibility and reuse of code.

Carlintyj committed:
Add assertions for input validation and error handling

This commit adds assertions to improve the input validation and error
handling in the code. The assertions ensure that user input is not null
or empty, that commands are correctly identified, and that indices and
task descriptions are within valid ranges and formats.

The assertions help catch potential issues early in development and
prevent unexpected behavior when processing user commands.

Carlintyj committed:
Merge pull request #1 from Carlintyj/branch-A-Assertions

Add assertions for input validation and error handling

Carlintyj committed:
Merge branch 'master' into branch-A-CodeQuality

Carlintyj committed:
Merge pull request #2 from Carlintyj/branch-A-CodeQuality

Command classes: complicated expressions and deep nestings.

Carlintyj committed:
Set up CI

Carlintyj committed:
Add the ability to recognize and deal with duplicate items

Carlintyj committed:
Update GUI of the chatbot to improve its user interface

Carlintyj committed:
Improve checkstyle in Duke, AddCommand, HelpCommand, Parser, TaskList,
DialogBox, MainWindow and Ui.

Carlintyj committed:
Updated readme for user guide

Carlintyj committed:
Update README.md

Carlintyj committed:
Fix minor bug fix on date time format not throwing an exception.

Carlintyj committed:
Merge branch 'master' of https://github.com/Carlintyj/ip

* 'master' of https://github.com/Carlintyj/ip:
  Update README.md

Carlintyj committed:
Update README.md
